// @any-touch/swipe/dist/index.es.js
import { __extends, __assign } from 'tslib';
import { INPUT_END } from '@any-touch/shared';
import { ComputeMaxLength, ComputeVAndDir, ComputeDistance } from '@any-touch/compute';
import Recognizer from '@any-touch/recognizer';

var DEFAULT_OPTIONS = {
    name: 'swipe',
    threshold: 10,
    velocity: 0.3,
    pointLength: 1,
};
var default_1 = (function (_super) {
    __extends(default_1, _super);
    function default_1(options) {
        return _super.call(this, __assign(__assign({}, DEFAULT_OPTIONS), options)) || this;
    }
    default_1.prototype.test = function (input) {
        var inputType = input.inputType;
        if (INPUT_END !== inputType)
            return false;
        var _a = this.computed, velocityX = _a.velocityX, velocityY = _a.velocityY, maxPointLength = _a.maxPointLength, distance = _a.distance;
        return this.options.pointLength === maxPointLength &&
            this.options.threshold < distance &&
            this.options.velocity < Math.max(velocityX, velocityY);
    };
    default_1.prototype.recognize = function (input, emit) {
        this.computed = this.compute([ComputeMaxLength, ComputeVAndDir, ComputeDistance], input);
        if (this.test(input)) {
            emit(this.options.name, this.computed);
            emit(this.options.name + this.computed.direction, this.computed);
        }
    };
    return default_1;
}(Recognizer));

export default default_1;
