// @any-touch/shared/dist/index.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ObjectToString = Object.prototype.toString;
function isRegExp(input) {
    return '[object RegExp]' === ObjectToString.call(input);
}
function isFunction(input) {
    return '[object Function]' === ObjectToString.call(input);
}

var CLIENT_X = 'clientX';
var CLIENT_Y = 'clientY';
var COMPUTE_INTERVAL = 16;
var INPUT_START = 'start';
var INPUT_MOVE = 'move';
var INPUT_CANCEL = 'cancel';
var INPUT_END = 'end';
var DIRECTION_LEFT = 'left';
var DIRECTION_RIGHT = 'right';
var DIRECTION_UP = 'up';
var DIRECTION_DOWN = 'down';
var NONE = 'none';
var TOUCH = 'touch';
var MOUSE = 'mouse';
var TOUCH_START = TOUCH + INPUT_START;
var TOUCH_MOVE = TOUCH + INPUT_MOVE;
var TOUCH_END = TOUCH + INPUT_END;
var TOUCH_CANCEL = TOUCH + INPUT_CANCEL;
var MOUSE_UP = MOUSE + DIRECTION_UP;
var MOUSE_MOVE = MOUSE + INPUT_MOVE;
var MOUSE_DOWN = MOUSE + DIRECTION_DOWN;
var SUPPORT_TOUCH = "on" + TOUCH_START in window;
var STATUS_POSSIBLE = 'p';
var STATUS_START = INPUT_START;
var STATUS_MOVE = INPUT_MOVE;
var STATUS_END = INPUT_END;
var STATUS_RECOGNIZED = 'r';
var STATUS_FAILED = 'f';
var STATUS_CANCELLED = INPUT_CANCEL;

function round2(n) {
    return Math.round(n * 100) / 100;
}

exports.CLIENT_X = CLIENT_X;
exports.CLIENT_Y = CLIENT_Y;
exports.COMPUTE_INTERVAL = COMPUTE_INTERVAL;
exports.DIRECTION_DOWN = DIRECTION_DOWN;
exports.DIRECTION_LEFT = DIRECTION_LEFT;
exports.DIRECTION_RIGHT = DIRECTION_RIGHT;
exports.DIRECTION_UP = DIRECTION_UP;
exports.INPUT_CANCEL = INPUT_CANCEL;
exports.INPUT_END = INPUT_END;
exports.INPUT_MOVE = INPUT_MOVE;
exports.INPUT_START = INPUT_START;
exports.MOUSE = MOUSE;
exports.MOUSE_DOWN = MOUSE_DOWN;
exports.MOUSE_MOVE = MOUSE_MOVE;
exports.MOUSE_UP = MOUSE_UP;
exports.NONE = NONE;
exports.STATUS_CANCELLED = STATUS_CANCELLED;
exports.STATUS_END = STATUS_END;
exports.STATUS_FAILED = STATUS_FAILED;
exports.STATUS_MOVE = STATUS_MOVE;
exports.STATUS_POSSIBLE = STATUS_POSSIBLE;
exports.STATUS_RECOGNIZED = STATUS_RECOGNIZED;
exports.STATUS_START = STATUS_START;
exports.SUPPORT_TOUCH = SUPPORT_TOUCH;
exports.TOUCH = TOUCH;
exports.TOUCH_CANCEL = TOUCH_CANCEL;
exports.TOUCH_END = TOUCH_END;
exports.TOUCH_MOVE = TOUCH_MOVE;
exports.TOUCH_START = TOUCH_START;
exports.isFunction = isFunction;
exports.isRegExp = isRegExp;
exports.round2 = round2;
