// @any-touch/rotate/dist/index.es.js
import { __extends, __assign } from 'tslib';
import { ComputeVectorForMutli, computeAngle } from '@any-touch/compute';
import Recognizer, { recognizeForPressMoveLike } from '@any-touch/recognizer';

var DEFAULT_OPTIONS = {
    name: 'rotate',
    threshold: 0,
    pointLength: 2,
};
var default_1 = (function (_super) {
    __extends(default_1, _super);
    function default_1(options) {
        return _super.call(this, __assign(__assign({}, DEFAULT_OPTIONS), options)) || this;
    }
    default_1.prototype.test = function (input) {
        var pointLength = input.pointLength;
        var angle = this.computed.angle;
        return this.isValidPointLength(pointLength) && (this.options.threshold < Math.abs(angle) || this.isRecognized);
    };
    default_1.prototype.recognize = function (input, emit) {
        var computed = this.compute([ComputeVectorForMutli], input);
        if ("activeV" in computed) {
            this.computed = __assign(__assign({}, this.computed), computeAngle(computed));
        }
        recognizeForPressMoveLike(this, input, emit);
    };
    return default_1;
}(Recognizer));

export default default_1;
