// @any-touch/recognizer/dist/index.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var shared = require('@any-touch/shared');

function resetStatus (recognizer) {
    if (-1 !== [shared.STATUS_END, shared.STATUS_CANCELLED, shared.STATUS_RECOGNIZED, shared.STATUS_FAILED].indexOf(recognizer.status)) {
        recognizer.status = shared.STATUS_POSSIBLE;
    }
}

function flow(isVaild, activeStatus, inputType) {
    var _a, _b, _c, _d, _e, _f, _g;
    var STATE_MAP = {
        1: (_a = {},
            _a[shared.STATUS_POSSIBLE] = (_b = {},
                _b[shared.INPUT_MOVE] = shared.STATUS_START,
                _b),
            _a[shared.STATUS_START] = (_c = {},
                _c[shared.INPUT_MOVE] = shared.STATUS_MOVE,
                _c[shared.INPUT_END] = shared.STATUS_END,
                _c[shared.INPUT_CANCEL] = shared.STATUS_CANCELLED,
                _c),
            _a[shared.STATUS_MOVE] = (_d = {},
                _d[shared.INPUT_MOVE] = shared.STATUS_MOVE,
                _d[shared.INPUT_END] = shared.STATUS_END,
                _d[shared.INPUT_CANCEL] = shared.STATUS_CANCELLED,
                _d),
            _a),
        0: (_e = {},
            _e[shared.STATUS_START] = (_f = {},
                _f[shared.INPUT_MOVE] = shared.STATUS_CANCELLED,
                _f[shared.INPUT_END] = shared.STATUS_END,
                _f[shared.INPUT_CANCEL] = shared.STATUS_CANCELLED,
                _f),
            _e[shared.STATUS_MOVE] = (_g = {},
                _g[shared.INPUT_START] = shared.STATUS_FAILED,
                _g[shared.INPUT_MOVE] = shared.STATUS_CANCELLED,
                _g[shared.INPUT_END] = shared.STATUS_END,
                _g[shared.INPUT_CANCEL] = shared.STATUS_CANCELLED,
                _g),
            _e)
    };
    if (void 0 !== STATE_MAP[Number(isVaild)][activeStatus]) {
        return STATE_MAP[Number(isVaild)][activeStatus][inputType] || activeStatus;
    }
    else {
        return activeStatus;
    }
}
function recognizeForPressMoveLike (recognizer, input, emit) {
    var isVaild = recognizer.test(input);
    resetStatus(recognizer);
    var inputType = input.inputType;
    recognizer.status = flow(isVaild, recognizer.status, inputType);
    var computed = recognizer.computed;
    recognizer.isRecognized = [shared.STATUS_START, shared.STATUS_MOVE].includes(recognizer.status);
    var name = recognizer.name, status = recognizer.status, isRecognized = recognizer.isRecognized;
    if (isRecognized) {
        emit(name, computed);
    }
    if (isRecognized || [shared.STATUS_END, shared.STATUS_CANCELLED].includes(recognizer.status)) {
        emit(name + status, computed);
    }
    return isVaild;
}

var default_1 = (function () {
    function default_1(options) {
        this.disabled = false;
        this.status = shared.STATUS_POSSIBLE;
        this.isRecognized = false;
        this.recognizerMap = {};
        this.computedGroup = {};
        this.computed = {};
        this.computeFunctionMap = {};
        this.options = options;
        this.name = this.options.name;
    }
    default_1.prototype.set = function (options) {
        if (void 0 !== options) {
            this.options = tslib.__assign(tslib.__assign({}, this.options), options);
        }
        return this;
    };
    default_1.prototype.isValidPointLength = function (pointLength) {
        return 0 === this.options.pointLength || this.options.pointLength === pointLength;
    };
    default_1.prototype.compute = function (Cs, input) {
        var e_1, _a;
        var computed = Object.create(null);
        try {
            for (var Cs_1 = tslib.__values(Cs), Cs_1_1 = Cs_1.next(); !Cs_1_1.done; Cs_1_1 = Cs_1.next()) {
                var C = Cs_1_1.value;
                var _id = C._id;
                var _b = this, computedGroup = _b.computedGroup, computeFunctionMap = _b.computeFunctionMap;
                if (void 0 === computeFunctionMap[_id]) {
                    computeFunctionMap[_id] = C();
                }
                computedGroup[_id] = computedGroup[_id] || computeFunctionMap[_id](input);
                for (var key in computedGroup[_id]) {
                    computed[key] = computedGroup[_id][key];
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (Cs_1_1 && !Cs_1_1.done && (_a = Cs_1.return)) _a.call(Cs_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return computed;
    };
    return default_1;
}());

exports.default = default_1;
exports.recognizeForPressMoveLike = recognizeForPressMoveLike;
exports.resetStatusForPressMoveLike = resetStatus;
